File: Daxelo-Kinrel-App/lib/core/kinship/v3/kinship_signature.dart
// lib/core/kinship/v3/kinship_signature.dart
//
// DAXELO KINREL — Deterministic Kinship Engine v3.0
// KinshipSignature — runtime-only structured data representing the
// relationship between two persons. NEVER stored. NEVER persisted.
//
// A signature is built from the BFS traversal path between Person A
// and Person B using only fundamental edge primitives:
//   UP_PARENT, DOWN_CHILD, SPOUSE, UP_ADOPTIVE_PARENT, UP_STEP_PARENT

/// The traversal primitives used by the engine. These are the ONLY
/// allowed edge types — derived relationships (brother, uncle, etc.)
/// are NEVER traversal primitives.
enum TraversePrimitive {
  upParent,        // child → parent (stored: parent A→B, traversed B→A)
  downChild,       // parent → child (stored: parent A→B, traversed A→B)
  spouse,          // bidirectional (stored: spouse A→B)
  upAdoptiveParent, // child → adoptive parent
  upStepParent,    // child → step-parent
}

/// Consanguinity type — determines whether the relationship is blood,
/// half, step, adoptive, or through marriage.
enum Consanguinity {
  blood,
  half,
  step,
  adoptive,
  inLaw,
}

/// Which side of the family the relationship belongs to.
enum FamilySide {
  paternal,
  maternal,
  none, // pure spouse traversal or no parent branch
}

/// A runtime-only structured representation of the kinship between
/// two persons. Built from the BFS path, consumed by the Vocabulary
/// Mapper. NEVER stored in the database.
class KinshipSignature {
  const KinshipSignature({
    required this.generationDelta,
    required this.pathPattern,
    required this.side,
    required this.consanguinity,
    required this.genderAnchor,
    required this.seniority,
    required this.removal,
    required this.doubleKinship,
  });

  /// Generation difference: negative = ancestor, positive = descendant,
  /// 0 = same generation. Range: -8 to +8.
  final int generationDelta;

  /// The canonical path pattern, e.g. "UP_PARENT_UP_PARENT" for
  /// grandparent. Built by joining TraversePrimitive names with _.
  final String pathPattern;

  /// Which side of the family (paternal/maternal/none).
  final FamilySide side;

  /// Blood / half / step / adoptive / inLaw.
  final Consanguinity consanguinity;

  /// Gender of the target person (male/female/neutral).
  final String genderAnchor;

  /// Birth order: elder / younger / twin / none.
  final String seniority;

  /// Removal count for "once removed" cousins.
  final int removal;

  /// True for double first cousins, etc.
  final bool doubleKinship;

  /// Builds the path pattern string from a list of primitives.
  static String buildPattern(List<TraversePrimitive> path) {
    if (path.isEmpty) return '';
    return path.map(_primitiveName).join('_');
  }

  static String _primitiveName(TraversePrimitive p) {
    switch (p) {
      case TraversePrimitive.upParent: return 'UP_PARENT';
      case TraversePrimitive.downChild: return 'DOWN_CHILD';
      case TraversePrimitive.spouse: return 'SPOUSE';
      case TraversePrimitive.upAdoptiveParent: return 'UP_ADOPTIVE_PARENT';
      case TraversePrimitive.upStepParent: return 'UP_STEP_PARENT';
    }
  }

  @override
  String toString() =>
      'KinshipSignature(gen=$generationDelta, pattern=$pathPattern, '
      'side=$side, consang=$consanguinity, gender=$genderAnchor, '
      'senior=$seniority, removal=$removal, double=$doubleKinship)';
}

File: kinrel-v3/server/src/modules/kinship/kinship-signature.ts
/**
 * Daxelo-Kinrel — KinshipSignature (spec §6)
 * ===========================================
 * Runtime-only structured data describing a kinship relationship.
 *
 * CRITICAL: This object is NEVER persisted to the database.
 * It is computed on-demand by GraphEngineService and consumed by
 * KinshipService to look up the localized term.
 *
 * The signatureKey is the deterministic lookup key for KinshipVocabulary.
 * Same graph + same Person A + same Person B = same signatureKey. Always.
 */

export type Side = "paternal" | "maternal" | "none";
export type Consanguinity = "blood" | "half" | "step" | "adoptive" | "inLaw" | "foster" | "spiritual";
export type GenderAnchor = "male" | "female" | "neutral";
export type Seniority = "elder" | "younger" | "twin" | "none";
export type Temporal = "current" | "former" | "late";

// Traversal primitives — only these are allowed (spec §5).
// Forbidden primitives: BROTHER, SISTER, UNCLE, AUNT, COUSIN, GRANDFATHER.
export type Primitive =
  | "UP_PARENT"
  | "DOWN_CHILD"
  | "SPOUSE"
  | "UP_ADOPTIVE_PARENT"
  | "DOWN_ADOPTIVE_CHILD"
  | "UP_STEP_PARENT"
  | "DOWN_STEP_CHILD"
  | "UP_FOSTER_PARENT"
  | "DOWN_FOSTER_CHILD"
  | "UP_SPIRITUAL_PARENT"
  | "DOWN_SPIRITUAL_CHILD";

export interface KinshipSignature {
  generationDelta: number;        // -8 .. +8
  pathPattern: string;            // primitives joined by "_"  (spec §6.1 grammar)
  side: Side;                     // spec §6.3 — set by first UP_PARENT
  consanguinity: Consanguinity;   // spec §6.4 — full / half / step / adoptive / inLaw / foster / spiritual
  genderAnchor: GenderAnchor;     // derived from target Person.gender
  seniority: Seniority;           // sibling birth order (spec §6.4)
  removal: number;                // cousin removal (spec §6.5)
  doubleKinship: boolean;         // double first cousins etc. (spec §6.5)
  temporal: Temporal;             // current | former | late (NEW v3.1 — spec §6)
}

export function signatureKey(s: KinshipSignature): string {
  return [
    `g=${s.generationDelta}`,
    `p=${s.pathPattern}`,
    `s=${s.side}`,
    `c=${s.consanguinity}`,
    `x=${s.genderAnchor}`,
    `r=${s.removal}`,
    `sn=${s.seniority}`,
    `d=${s.doubleKinship ? 1 : 0}`,
    `t=${s.temporal}`,
  ].join("|");
}

// ---------------------------------------------------------------------------
// Path pattern utilities (spec §6.1)
// ---------------------------------------------------------------------------

export function buildPathPattern(primitives: Primitive[]): string {
  return primitives.join("_");
}

export function parsePathPattern(pattern: string): Primitive[] {
  return pattern.split("_").filter(Boolean) as Primitive[];
}

/**
 * Count UP vs DOWN primitives to compute generationDelta (spec §6).
 * Each UP_PARENT decrements gen, each DOWN_CHILD increments.
 * SPOUSE is gen-neutral.
 */
export function computeGenerationDelta(pattern: string): number {
  const primitives = parsePathPattern(pattern);
  let delta = 0;
  for (const p of primitives) {
    if (p.startsWith("UP_")) delta -= 1;
    else if (p.startsWith("DOWN_")) delta += 1;
    // SPOUSE → 0
  }
  return delta;
}

/**
 * Determine side from path pattern (spec §6.3).
 * Side is set by the FIRST UP_PARENT (or UP_ADOPTIVE_PARENT / UP_STEP_PARENT).
 * The side paternal/maternal is then resolved at runtime by looking at
 * the gender of the parent at that step.
 *
 * This function returns the LITERAL side token only when known from the
 * pattern alone; in most cases the engine must consult the graph.
 */
export function inferSideFromPattern(pattern: string): Side {
  if (!pattern.includes("UP_")) return "none";
  // We cannot resolve paternal vs maternal from the pattern alone.
  // The engine must set this by inspecting the actual parent's gender
  // at the first UP step.
  return "none"; // engine overrides
}
